### ISSUE TITLE
UndefinedRecorder with Jinja loop issue

### ISSUE THREAD

**Issue by rogalski** [sqlfluff/sqlfluff#6360]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.


### What Happened

Unrecoverable failure.

### Expected Behaviour

Templating error.

### Observed Behaviour

Unrecoverable failure.

### How to reproduce

repro1.sql
```sql
SELECT a FROM {{ tbl }};
```

repro2.sql
```sql
{% for tbl in tbl_list %}
SELECT a FROM {{ tbl }};
{% endfor %}
```

```
sqlfluff lint -d ansi repro1.sql repro2.sql 
== [repro1.sql] FAIL                                                                                                                                                                                                         
L:   1 | P:  10 |  PRS | Line 1, Position 10: Found unparsable section: 'FROM'                                                                                                                                               
L:   1 | P:  14 | CV06 | Statements must end with a semi-colon.
                       | [convention.terminator]
L:   1 | P:  18 |  TMP | Undefined jinja template variable: 'tbl'
L:   1 | P:  24 | LT12 | Files must end with a single trailing newline.
                       | [layout.end_of_file]
WARNING: Parsing errors found and dialect is set to 'ansi'. Have you configured your dialect correctly?                                                                                                                      
== [repro2.sql] FAIL                                                                                                                                                                                                         
L:   1 | P:   1 |  TMP | Unrecoverable failure in Jinja templating:
                       | 'UndefinedRecorder' object is not iterable. Have you
                       | correctly configured your variables?
                       | https://docs.sqlfluff.com/en/latest/perma/variables.html
All Finished 📜 🎉!                                                          
```

### Dialect

ansi

### Version

3.2.4

### Configuration

not applicable

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)
